src/interface/meshblock_data.cpp

```diff
@@ -53,6 +53,12 @@ void MeshBlockData<T>::Initialize(
   for (auto const &q : resolved_packages->AllFields()) {
     AddField(q.first.base_name, q.second, q.first.sparse_id);
   }
+
+  Metadata::FlagCollection flags({Metadata::Sparse, Metadata::ForceAllocOnNewBlocks});
+  auto vars = GetVariablesByFlag(flags);
+  for (auto &v : vars.vars()) {
+    AllocateSparse(v->label());
+  }
```

Comment:
As it's written right now, `Metadata::ForceAllocOnNewBlocks` has no effect on any variable that doesn't have `Metadata::Sparse`.  Those fields will always be allocated anyway, of course.  I could add some kind of check that'll throw if you set `Metadata::ForceAllocOnNewBlocks` for a dense variable, but I didn't because you still get the behavior you expect.